admin/server/auth/middleware.go

```diff
@@ -0,0 +1,142 @@
+func (a *Authenticator) UnaryServerInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		authHeader := metautils.ExtractIncoming(ctx).Get("authorization")
+		newCtx, err := a.parseClaimsFromBearer(ctx, authHeader)
+		if err != nil {
+			return nil, status.Error(codes.Unauthenticated, err.Error())
+		}
+
+		return handler(newCtx, req)
+	}
+}
+
+// StreamServerInterceptor is the streaming variant of UnaryServerInterceptor.
+func (a *Authenticator) StreamServerInterceptor() grpc.StreamServerInterceptor {
+	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		authHeader := metautils.ExtractIncoming(ss.Context()).Get("authorization")
+		newCtx, err := a.parseClaimsFromBearer(ss.Context(), authHeader)
+		if err != nil {
+			return status.Error(codes.Unauthenticated, err.Error())
+		}
+
+		wrapped := grpc_middleware.WrapServerStream(ss)
+		wrapped.WrappedContext = newCtx
+
+		return handler(srv, wrapped)
+	}
+}
+
+// HTTPMiddleware is a HTTP middleware variant of UnaryServerInterceptor.
+// It additionally supports reading access tokens from cookies.
+// It should be used for non-gRPC HTTP endpoints (CookieAuthAnnotator takes care of handling cookies in gRPC-gateway requests).
+func (a *Authenticator) HTTPMiddleware(next gateway.HandlerFunc) gateway.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+		// Handle authorization header
+		authHeader := r.Header.Get("Authorization")
+		if authHeader != "" {
+			newCtx, err := a.parseClaimsFromBearer(r.Context(), authHeader)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusUnauthorized)
+				return
+			}
+
+			next(w, r.WithContext(newCtx), pathParams)
+			return
+		}
+
+		// There was no authorization header. Try the cookie.
+		sess, err := a.cookies.Get(r, cookieName)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to get session: %s", err), http.StatusInternalServerError)
+			return
+		}
+
+		// Read access token from cookie
+		authToken, ok := sess.Values[cookieFieldAccessToken].(string)
+		if ok && authToken != "" {
+			newCtx, err := a.parseClaimsFromToken(r.Context(), authToken)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusUnauthorized)
+				return
+			}
+
+			next(w, r.WithContext(newCtx), pathParams)
+			return
+		}
+
+		// No token was found. Set anonClaims.
+		newCtx := context.WithValue(r.Context(), claimsContextKey{}, anonClaims{})
+		next(w, r.WithContext(newCtx), pathParams)
+	}
+}
+
+func (a *Authenticator) parseClaimsFromBearer(ctx context.Context, authorizationHeader string) (context.Context, error) {
+	// If authorization header is not set, we set anonClaims.
+	if authorizationHeader == "" {
```

Comment:
There won't be – instead, the `anonClaims` will be set, which will basically deny any action. So the rejection in case of no auth happens in the handlers, not here.

This enables certain endpoints to be public, like health check, ping, and maybe enabling public dashboards.

Reply:
how does this work for local rill-developer, what type of `Claims` object will be set and where

Reply:
I meant for development

Reply:
summarizing discussion we had here - I think we should think of how we can make it simple for running the server in dev env without having dependency on an identity provider.